kernel/posix/mqueue.c

```diff
@@ -0,0 +1,532 @@
+	}
+
+	if (mqd->flags & O_NONBLOCK) {
+		timeout = K_NO_WAIT;
+	}
+
+	if (k_msgq_get(&msg_queue->queue, (void *)msg_ptr, timeout) == 0) {
+		ret = msg_queue->queue.msg_size;
+	} else {
+		errno = EAGAIN;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Receive message from a message queue within abstime time.
+ *
+ * See IEEE 1003.1
+ */
+int mq_timedreceive(mqd_t mqdes, char *msg_ptr, size_t msg_len,
+			unsigned int *msg_prio, const struct timespec *abstime)
+{
+	mqueue_desc *mqd = (mqueue_desc *)mqdes;
+	mqueue_object  *msg_queue = mqd->mqueue;
+	s32_t  timeout = K_NO_WAIT, ret = -1;
+
+	if (mqd == NULL) {
+		errno = EBADF;
+		return ret;
+	}
+
+	if (msg_len < msg_queue->queue.msg_size) {
+		errno = EAGAIN;
+		return ret;
+	}
+
+	if (!(mqd->flags & O_NONBLOCK)) {
+		timeout = (s32_t) calculate_timeout(abstime);
+	}
+
+	if (k_msgq_get(&msg_queue->queue, (void *)msg_ptr, timeout) == 0) {
+		ret = msg_queue->queue.msg_size;
+	} else {
+		errno = EAGAIN;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Send message to a message queue within abstime time.
+ *
+ * See IEEE 1003.1
+ */
+int mq_timedsend(mqd_t mqdes, const char *msg_ptr, size_t msg_len,
+		 unsigned int msg_prio, const struct timespec *abstime)
+{
+	mqueue_desc *mqd = (mqueue_desc *)mqdes;
+	mqueue_object  *msg_queue = mqd->mqueue;
+	s32_t  timeout = K_NO_WAIT, ret = -1;
+
+	if (mqd == NULL) {
+		errno = EBADF;
+		return -1;
+	}
+
+	if (msg_len < msg_queue->queue.msg_size) {
+		errno = EAGAIN;
+		return ret;
+	}
+
+	if (!(mqd->flags & O_NONBLOCK)) {
+		timeout = (s32_t) calculate_timeout(abstime);
+	}
+
+	if (k_msgq_put(&msg_queue->queue, (void *)msg_ptr, timeout) != 0) {
+		errno = EAGAIN;
+	} else {
+		ret = 0;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Get message queue attributes.
+ *
+ * See IEEE 1003.1
+ */
+int mq_getattr(mqd_t mqdes, struct mq_attr *mqstat)
+{
+	mqueue_desc *mqd = (mqueue_desc *)mqdes;
+	mqueue_object  *msg_queue = mqd->mqueue;
+	struct k_msgq_attrs attrs;
+	s32_t ret = 0;
+
+	if (mqd == NULL) {
+		errno = EBADF;
+		return -1;
+	}
+
+	k_sem_take(&mq_sem, K_FOREVER);
+	if (k_msgq_get_attrs(&msg_queue->queue, &attrs) == 0) {
+		mqstat->mq_flags = mqd->flags;
+		mqstat->mq_maxmsg = attrs.max_msgs;
+		mqstat->mq_msgsize = attrs.msg_size;
+		mqstat->mq_curmsgs = attrs.used_msgs;
+	} else {
+		ret = -1;
+	}
+
+	k_sem_give(&mq_sem);
+	return ret;
+}
+
+/**
+ * @brief Set message queue attributes.
+ *
+ * See IEEE 1003.1
+ */
+int mq_setattr(mqd_t mqdes, const struct mq_attr *mqstat,
+	       struct mq_attr *omqstat)
+{
+	mqueue_desc *mqd = (mqueue_desc *)mqdes;
+
+	if (mqd == NULL) {
+		errno = EBADF;
+		return -1;
+	}
+
+	if (mqstat->mq_flags != 0 && mqstat->mq_flags != O_NONBLOCK) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	if (omqstat != NULL) {
+		mq_getattr(mqdes, omqstat);
+	}
+
+	k_sem_take(&mq_sem, K_FOREVER);
+	mqd->flags = mqstat->mq_flags;
+	k_sem_give(&mq_sem);
+
+	return 0;
+}
+
+
+/* Internal functions */
+static mqueue_object *find_in_list(const char *name)
```

Comment:
Use an slist here.

Reply:
Yes we can manage with single linked list only. Sure I will change this to single linked list.